integration/upload_test.go

```diff
@@ -0,0 +1,243 @@
+	}
+
+	suite.T().Logf("test config: %+v", *cfg)
+
+	opts := mqtt.NewClientOptions().
+		AddBroker(cfg.Broker).
+		SetClientID(uuid.New().String()).
+		SetKeepAlive(30 * time.Second).
+		SetCleanSession(true).
+		SetAutoReconnect(true)
+
+	mqttClient := mqtt.NewClient(opts)
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		require.NoError(suite.T(), token.Error(), "connect to MQTT broker")
+	}
+
+	thingCfg, err := getThingConfig(mqttClient)
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "get thing config")
+	}
+
+	suite.T().Logf("thing config: %+v", *thingCfg)
+
+	dittoClient, err := ditto.NewClientMqtt(mqttClient, ditto.NewConfiguration())
+	if err == nil {
+		err = dittoClient.Connect()
+	}
+
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "initialize ditto client")
+	}
+
+	suite.dittoClient = dittoClient
+	suite.mqttClient = mqttClient
+	suite.cfg = cfg
+	suite.thingCfg = thingCfg
+
+	suite.thingURL = fmt.Sprintf("%s/api/2/things/%s", strings.TrimSuffix(cfg.DittoAddress, "/"), thingCfg.DeviceID)
+	suite.featureURL = fmt.Sprintf("%s/features/%s", suite.thingURL, featureID)
+	suite.checkUploadDir()
+}
+
+func (suite *testSuite) TearDownSuite() {
+	suite.dittoClient.Disconnect()
+	suite.mqttClient.Disconnect(uint(suite.cfg.MqttQuiesceMs))
+}
+
+func TestFileUpload(t *testing.T) {
+	suite.Run(t, new(testSuite))
+}
+
+func (suite *testSuite) TestHTTPUpload() {
+	if len(suite.cfg.HTTPServer) == 0 {
+		suite.T().Skip("HTTP_SERVER variable must be set")
+	}
+	upload := newHTTPUpload(suite.cfg.HTTPServer)
+	suite.testUpload(upload)
+}
+
+func (suite *testSuite) TestAzureUpload() {
+	options, err := uploaders.GetAzureTestOptions(suite.T())
+	require.Nil(suite.T(), err, "error getting azure test options")
+	upload := newAzureUpload(options)
+	suite.testUpload(upload)
+}
+
+func (suite *testSuite) TestAWSUpload() {
+	options := uploaders.GetAWSTestOptions(suite.T())
+	upload, err := newAWSUpload(options)
+	require.Nil(suite.T(), err, "error creating AWS client")
+	suite.testUpload(upload)
+}
+
+func (suite *testSuite) testUpload(testUpload upload) {
+	files, err := createTestFiles(suite.cfg.UploadDir)
+	defer deleteTestFiles(files)
+	require.Nil(suite.T(), err, "creating test files failed")
+
+	type uploadRequest struct {
+		CorrelationID string            `json:"correlationId"`
+		Options       map[string]string `json:"options"`
+	}
+	requestPath := fmt.Sprintf("/features/%s/outbox/messages/request", featureID)
+	filePaths := make(map[string]string)
+	triggerCh := make(chan bool)
+	conn, _ := suite.startEventListener(typeMessages, func(props map[string]interface{}) bool {
+		if requestPath == props["path"] {
+			if value, ok := props["value"]; ok {
+				jsonValue, err := json.Marshal(value)
+				if err != nil {
+					return false // skip
```

Comment:
Is it ok jusjt to ignore it? That's the value in file upload path. The value should be valid, if not, it has to be indicated.